### PR TITLE
Handle multiple imports in one use statement

### LINKED ISSUE #563: Handle multiple imports in one use statement
It'd be nice to be able to do imports like this:

```rust
use wallet_abi::{RequestFormat, WalletABI};
```

Currently, we can only either import _one_ item or _all_ items via single item imports or blob imports.

### PR DESCRIPTION
**Pull request by m-fawaz**
Closes #563 
This enables a wider range of cases than the one mentioned in the issue. Specifically, the target is to eventually support all patterns supported by Rust: https://doc.rust-lang.org/reference/items/use-declarations.html

Right now, the following should work:
`use a::b::{c}`
`use a::b::{c, d}`
`use a::b::{c as abc, d as abd}`
`use a::b::{c as abc, d::{e, f::g as abdfg}}`
`use a::b::{c as abc, d::{e, f::g, *}}`
and so on.
Using `Self` is not yet covered.

The parsing algorithm is recurisve since a list of items could have its own list of items etc.

I still need to add a comprehensive test but here I'm converting some of the existing tests to use the newly supported method which gives extra coverage.

Edit: Added a test that covers a few cases. The remaining gaps, I believe, are:
* Supporting `self` (as in: `use a::b::{self as ab, c as abc};`)
* Handling the case where we have a trailing `as` after the `}`. Rust erros out for this. e.g. `use a::{b, c} as d` (this, of course, is meanignless).

Edit 2: Two more changes:
* Handling trailing `as` after the `}`.
* Handling non-final `*` more generically in the language file.
So, the following should give a parsing error:
`use a::b::*::c`
`use a::b::c::{d, e::*::f}`
`use a::b::{c, d} as e`

### CODE REVIEW

**Comment by adlerjohn** [comment]
CI is failing for code formatting format. Rust has an amazing [canonical formatter](https://github.com/rust-lang/rustfmt) that you can use with `cargo fmt`.

**Comment by m-fawaz** [comment]
>

**Comment by m-fawaz** [comment]
@adlerjohn @sezna quick question: I don't see the test I added (`multi_item_import`) running here in CI: https://github.com/FuelLabs/sway/runs/4722647594?check_suite_focus=true - is there something I'm missing?

**Comment by adlerjohn** [comment]
It's running here actually (blame GitHub actions for being opaque): https://github.com/FuelLabs/sway/runs/4722647594?check_suite_focus=true#step:11:775

**Comment by m-fawaz** [comment]
> It's running here actually (blame GitHub actions for being opaque): https://github.com/FuelLabs/sway/runs/4722647594?check_suite_focus=true#step:11:775

Oh that's great! - I swear I looked there but couldn't find anything. I was relying on my browser to search through the page but it wasn't returning anything. I should've done a manual inspection instead. Thanks John!

**Comment by sezna** [comment]
You can also verify the tests run by running `cargo run` in the test directory. The only requirement is that `fuel-core` must be running.

**Comment by adlerjohn** [comment]
> I was relying on my browser to search through the page but it wasn't returning anything. I should've done a manual inspection instead.

Yeah I tried control+f at first and it didn't return anything. You actually have to go to the top-right and search through that box 🤷

**Comment by m-fawaz** [comment]
> You can also verify the tests run by running `cargo run` in the test directory. The only requirement is that `fuel-core` must be running.
I ran the tests locally before committing and they were okay! I just thought maybe Clippy is using an older version of the test directory or something like that (which in retrospect doens't make much sense).

**Comment by sezna** [comment]
Is this intended to be a draft or should we review it?

**Comment by m-fawaz** [comment]
> Is this intended to be a draft or should we review it?

This is ready for review now. I made additional changes. The only remaining gap is supporting `Self`.

**Comment by adlerjohn** [comment]
> The only remaining gap is supporting `Self`.

Can you file an issue describing the missing feature? Implementing in can be left to a subsequent PR.

**Comment by m-fawaz** [comment]
> > The only remaining gap is supporting `Self`.
> 
> Can you file an issue describing the missing feature? Implementing in can be left to a subsequent PR.

Done: https://github.com/FuelLabs/sway/issues/589

File: sway-core/src/parse_tree/use_statement.rs [hunk 1]
```diff
@@ -38,62 +37,136 @@ impl UseStatement {
...
+            warnings,
+            errors
+        );
+        top_level_alias = Some(alias_parsed);
+        last_item = import_path_vec.pop().unwrap();
+    }
+
+    for item in import_path_vec.into_iter() {
+        if item.as_rule() == Rule::ident {
+            import_path_buf.push(check!(
```
Comment (m-fawaz) [inline sway-core/src/parse_tree/use_statement.rs:80]: This is now guaranteed by the language actually. Every `item` should be an `ident`. There is probably no need for the `if`.

**Comment by m-fawaz** [comment]
I'm not sure why CI is failing. I can't reproduce the problem locally. Is this an OOM issue like this test https://github.com/FuelLabs/sway/blob/c55f7e76b010f9a0b7abcac2b7cfd6155a4236a7/test/src/e2e_vm_tests/mod.rs#L18 ? My test is very similar to it.

**Comment by sezna** [comment]
This looks like the VM reverted, which is what it does in the case of a panic, which is different. Not sure why this one is hitting a VM panic. 

But yeah, in general, we have a memory overhead issue when declaring dependencies. Minimizing the number of `dep` statements should help. 

I'll take a look at why you're hitting a REVERT. It's not immediately clear to me.

**Comment by m-fawaz** [comment]
> This looks like the VM reverted, which is what it does in the case of a panic, which is different. Not sure why this one is hitting a VM panic.
> 
> But yeah, in general, we have a memory overhead issue when declaring dependencies. Minimizing the number of `dep` statements should help.
> 
> I'll take a look at why you're hitting a REVERT. It's not immediately clear to me.

I needed to make one more change so CI will now run again. We'll see if the problem persists.

**Comment by m-fawaz** [comment]
I still cannot reproduce the CI test failure locally. I tried running the test individually as well as running the full suite (while `fuel-core` is running). So, in my last commit, I reduced the size of the test to see if that helps.

**Comment by adlerjohn** [comment]
cc @sezna for review before merge conflicts happen 😂

**Comment by sezna** [comment]
Really weird that the VM was hitting a revert, but I guess if that fixed it...I'll keep this in mind if we see that happen again

**Review by sezna — APPROVED** [review]

**Pull request merged**